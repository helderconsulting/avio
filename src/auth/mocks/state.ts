import type { MiddlewareHandler } from 'hono';
import type { AuthContext } from '../context.js';
import { UnauthorizedError } from '../error.js';
import { MockAuthService } from './service.js';

export const createMockAuthState: MiddlewareHandler<AuthContext> = async (
  c,
  next
) => {
  const authService = new MockAuthService();
  c.set('authService', authService);
  await next();
};

export const authenticated: MiddlewareHandler<AuthContext> = async (
  c,
  next
) => {
  await next();
};

export const unauthenticated: MiddlewareHandler<AuthContext> = () => {
  return Promise.reject(new UnauthorizedError());
};
