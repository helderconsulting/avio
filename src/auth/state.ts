import type { MiddlewareHandler } from 'hono';
import { createAuth } from '../lib/auth.js';
import type { AuthContext } from './context.js';
import { AuthService } from './service.js';

export const createAuthState: MiddlewareHandler<AuthContext> = async (
  c,
  next
) => {
  const db = c.get('db');
  const auth = createAuth(db);

  const authService = new AuthService(auth);
  c.set('authService', authService);

  await next();
};

export const authGuard: MiddlewareHandler<AuthContext> = async (c, next) => {
  const service = c.get('authService');
  await service.whoAmI(c.req.raw.headers);
  await next();
};
