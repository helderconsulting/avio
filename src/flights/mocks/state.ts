import type { MiddlewareHandler } from 'hono';
import type { WithId } from 'mongodb';
import type { FlightsContext } from '../context.js';
import type { FlightRequest } from '../schema.js';
import { MockFlightsService } from './service.js';

export const createCollection = (flights: WithId<FlightRequest>[]) => {
  const collection = new Map<string, WithId<FlightRequest>>();
  flights.forEach((flight) => {
    collection.set(flight._id.toHexString(), flight);
  });
  return collection;
};

export const createMockService = (flights: WithId<FlightRequest>[]) => {
  const collection = createCollection(flights);
  return new MockFlightsService(collection);
};

export const createMockFlightsState =
  (flights: WithId<FlightRequest>[]): MiddlewareHandler<FlightsContext> =>
  async (c, next) => {
    const mockService = createMockService(flights);
    c.set('flightsService', mockService);
    await next();
  };
