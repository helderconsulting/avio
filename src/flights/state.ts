import type { MiddlewareHandler } from 'hono';
import type { Db } from 'mongodb';
import type { FlightsContext } from './context.js';
import type { FlightEntity } from './schema.js';
import { FlightsService } from './service.js';

export const createFlightsService = (db: Db) => {
  const collection = db.collection<FlightEntity>('flights');
  return new FlightsService(collection);
};

export const createFlightsState: MiddlewareHandler<FlightsContext> = async (
  c,
  next
) => {
  const db = c.get('db');
  const flightsService = createFlightsService(db);
  c.set('flightsService', flightsService);

  await next();
};
