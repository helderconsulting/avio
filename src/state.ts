import type { Context, MiddlewareHandler } from 'hono';
import type { AppContext, PartialAppContext } from './context.js';
import { AppError } from './error.js';
import type { ConnectionInterface } from './lib/types.js';

export const createAppState =
  (connection: ConnectionInterface): MiddlewareHandler<AppContext> =>
  async (c, next) => {
    const { logger } = c.var;
    try {
      const state = c as Context<PartialAppContext>;
      logger.debug('Creating App State');

      if (!state.get('db')) {
        const db = await connection.connect();
        logger.debug(
          { databaseName: db.databaseName, options: db.options },
          'Database connected'
        );
        state.set('db', db);
      }

      logger.debug({ state: Object.keys(c.var) }, 'App State created');
      await next();
    } catch (error) {
      logger.error(error);
      throw new AppError();
    }
  };
