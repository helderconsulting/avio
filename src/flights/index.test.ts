import type { Hono } from 'hono';
import { ObjectId } from 'mongodb';
import type { WithId } from 'mongodb';
import { authenticated, unauthenticated } from '../auth/mocks/state.js';
import type { FlightsContext } from './context.js';
import { createMockFlightsState } from './mocks/state.js';
import type { FlightRequest } from './schema.js';
import { createRouter } from './index.js';

const validPayload = JSON.stringify({
  aircraft: 'CSTRC',
  flightNumber: 'AVIO205',
  schedule: {
    std: '2025-06-05T22:00:00.000Z',
    sta: '2025-06-05T22:00:00.000Z',
  },
  departure: 'LPPD',
  destination: 'LPLA',
});

const invalidPayload = JSON.stringify({
  aircraft: 'CSTRC',
  flightNumber: 'AVIO205',
  schedule: {
    std: '2024-09-30T22:00:00.000Z',
    sta: '2024-09-30T23:00:00.000Z',
  },
  departure: 'INCORRECT_DEPARTURE_LENGTH',
  destination: 'LPLA',
});

describe('Flights Endpoint', () => {
  describe('GET /flights', () => {
    const flightId = '6841cade4cace03b8f75235b';
    let authenticatedRouter: Hono<FlightsContext>;
    let unauthenticatedRouter: Hono<FlightsContext>;
    beforeAll(() => {
      const collection = [
        {
          _id: new ObjectId(flightId),
          aircraft: 'CSTRC',
          flightNumber: 'AVIO205',
          schedule: {
            std: '2024-09-30T22:00:00.000Z',
            sta: '2024-09-30T23:00:00.000Z',
          },
          departure: 'LPPD',
          destination: 'LPLA',
        },
      ];
      authenticatedRouter = createRouter(
        createMockFlightsState(collection),
        authenticated
      );
      unauthenticatedRouter = createRouter(
        createMockFlightsState(collection),
        unauthenticated
      );
    });

    test('should return a list of flights', async () => {
      const result = await authenticatedRouter.request('/');

      const json = await result.json<WithId<FlightRequest>[]>();

      expect(json).toMatchSnapshot();
      expect(result.status).toBe(200);
    });

    test('should return 401 if not authenticated', async () => {
      const result = await unauthenticatedRouter.request('/');

      expect(result.status).toBe(401);
    });

    test('should return 500 on server error', async () => {
      // ...
    });
  });

  describe('POST /flights', () => {
    let authenticatedRouter: Hono<FlightsContext>;
    let unauthenticatedRouter: Hono<FlightsContext>;
    beforeAll(() => {
      const collection: WithId<FlightRequest>[] = [];
      authenticatedRouter = createRouter(
        createMockFlightsState(collection),
        authenticated
      );
      unauthenticatedRouter = createRouter(
        createMockFlightsState(collection),
        unauthenticated
      );
    });
    test('should create a flight with valid data', async () => {
      const result = await authenticatedRouter.request('/', {
        method: 'POST',
        body: validPayload,
        headers: new Headers({ 'Content-Type': 'application/json' }),
      });

      const json = await result.json<WithId<FlightRequest>>();

      expect(json).toMatchSnapshot();
      expect(result.status).toBe(201);
    });

    test('should return 400 for invalid payload', async () => {
      const result = await authenticatedRouter.request('/', {
        method: 'POST',
        body: invalidPayload,
        headers: new Headers({ 'Content-Type': 'application/json' }),
      });

      const text = await result.text();

      expect(text).toBe('Invalid payload');
      expect(result.status).toBe(400);
    });

    test('should return 401 if not authenticated', async () => {
      const result = await unauthenticatedRouter.request('/', {
        method: 'POST',
        body: validPayload,
        headers: new Headers({ 'Content-Type': 'application/json' }),
      });

      expect(result.status).toBe(401);
    });

    test('should return 500 on server error', async () => {
      // ...
    });
  });

  describe('GET /flights/:flightId', () => {
    const flightId = '6841cade4cace03b8f75235b';
    let authenticatedRouter: Hono<FlightsContext>;
    let unauthenticatedRouter: Hono<FlightsContext>;
    beforeAll(() => {
      const collection = [
        {
          _id: new ObjectId(flightId),
          aircraft: 'CSTRC',
          flightNumber: 'AVIO205',
          schedule: {
            std: '2024-09-30T22:00:00.000Z',
            sta: '2024-09-30T23:00:00.000Z',
          },
          departure: 'LPPD',
          destination: 'LPLA',
        },
      ];
      authenticatedRouter = createRouter(
        createMockFlightsState(collection),
        authenticated
      );
      unauthenticatedRouter = createRouter(
        createMockFlightsState(collection),
        unauthenticated
      );
    });
    test('should return a flight by id', async () => {
      const result = await authenticatedRouter.request(`/${flightId}`);

      const json = await result.json();

      expect(json).toMatchSnapshot();
      expect(result.status).toBe(200);
    });

    test('should return 401 if not authenticated', async () => {
      const result = await unauthenticatedRouter.request(`/${flightId}`);
      expect(result.status).toBe(401);
    });

    test('should return 404 if flight is not found', async () => {
      const result = await authenticatedRouter.request(
        '/6841cade4cace03b8f75235d'
      );
      expect(result.status).toBe(404);
    });

    test('should return 500 on server error', async () => {
      // ...
    });
  });

  describe('PATCH /flights/:flightId', () => {
    const flightId = '6841cade4cace03b8f75235b';
    let authenticatedRouter: Hono<FlightsContext>;
    let unauthenticatedRouter: Hono<FlightsContext>;
    beforeAll(() => {
      const collection = [
        {
          _id: new ObjectId(flightId),
          aircraft: 'CSTRC',
          flightNumber: 'AVIO205',
          schedule: {
            std: '2024-09-30T22:00:00.000Z',
            sta: '2024-09-30T23:00:00.000Z',
          },
          departure: 'LPPD',
          destination: 'LPLA',
        },
      ];
      authenticatedRouter = createRouter(
        createMockFlightsState(collection),
        authenticated
      );
      unauthenticatedRouter = createRouter(
        createMockFlightsState(collection),
        unauthenticated
      );
    });
    test('should update a flight', async () => {
      const result = await authenticatedRouter.request(`/${flightId}`, {
        method: 'PATCH',
        body: validPayload,
        headers: new Headers({ 'Content-Type': 'application/json' }),
      });

      const json = await result.json();

      expect(json).toMatchSnapshot();
      expect(result.status).toBe(200);
    });

    test('should return 400 for invalid payload', async () => {
      const result = await authenticatedRouter.request(`/${flightId}`, {
        method: 'PATCH',
        body: invalidPayload,
        headers: new Headers({ 'Content-Type': 'application/json' }),
      });
      expect(result.status).toBe(400);
    });

    test('should return 401 if not authenticated', async () => {
      const result = await unauthenticatedRouter.request(`/${flightId}`, {
        method: 'PATCH',
        body: validPayload,
        headers: new Headers({ 'Content-Type': 'application/json' }),
      });
      expect(result.status).toBe(401);
    });

    test('should return 404 if flight is not found', async () => {
      const objectId = new ObjectId();
      const result = await authenticatedRouter.request(
        `/${objectId.toHexString()}`,
        {
          method: 'PATCH',
          body: validPayload,
          headers: new Headers({ 'Content-Type': 'application/json' }),
        }
      );
      expect(result.status).toBe(404);
    });

    test('should return 500 on server error', async () => {
      // ...
    });
  });
});

describe('DELETE /flights/:flightId', () => {
  const flightId = '6841cade4cace03b8f75235b';
  let authenticatedRouter: Hono<FlightsContext>;
  let unauthenticatedRouter: Hono<FlightsContext>;
  beforeAll(() => {
    const collection = [
      {
        _id: new ObjectId(flightId),
        aircraft: 'CSTRC',
        flightNumber: 'AVIO205',
        schedule: {
          std: '2024-09-30T22:00:00.000Z',
          sta: '2024-09-30T23:00:00.000Z',
        },
        departure: 'LPPD',
        destination: 'LPLA',
      },
    ];
    authenticatedRouter = createRouter(
      createMockFlightsState(collection),
      authenticated
    );
    unauthenticatedRouter = createRouter(
      createMockFlightsState(collection),
      unauthenticated
    );
  });
  test('should delete a flight by id', async () => {
    const result = await authenticatedRouter.request(`/${flightId}`, {
      method: 'DELETE',
    });
    expect(result.status).toBe(204);
  });

  test('should return 401 if not authenticated', async () => {
    const result = await unauthenticatedRouter.request(`/${flightId}`, {
      method: 'DELETE',
    });
    expect(result.status).toBe(401);
  });

  test('should return 404 if flight is not found', async () => {
    const objectId = new ObjectId();
    const result = await authenticatedRouter.request(
      `/${objectId.toHexString()}`,
      {
        method: 'DELETE',
      }
    );
    expect(result.status).toBe(404);
  });

  test('should return 500 on server error', async () => {
    // ...
  });
});
