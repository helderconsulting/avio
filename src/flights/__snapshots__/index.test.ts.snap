// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Flights Endpoint > GET /flights > should return a list of flights 1`] = `
[
  {
    "aircraft": "CSTRC",
    "departure": "LPPD",
    "destination": "LPLA",
    "flightNumber": "AVIO205",
    "id": "6841cade4cace03b8f75235b",
    "schedule": {
      "sta": "2024-09-30T23:00:00.000Z",
      "std": "2024-09-30T22:00:00.000Z",
    },
  },
]
`;

exports[`Flights Endpoint > GET /flights/:flightId > should return a flight by id 1`] = `
{
  "aircraft": "CSTRC",
  "departure": "LPPD",
  "destination": "LPLA",
  "flightNumber": "AVIO205",
  "id": "6841cade4cace03b8f75235b",
  "schedule": {
    "sta": "2024-09-30T23:00:00.000Z",
    "std": "2024-09-30T22:00:00.000Z",
  },
}
`;

exports[`Flights Endpoint > PATCH /flights/:flightId > should update a flight 1`] = `
{
  "aircraft": "CSTRC",
  "departure": "LPPD",
  "destination": "LPLA",
  "flightNumber": "AVIO205",
  "id": "6841cade4cace03b8f75235b",
  "schedule": {
    "sta": "2025-06-05T22:00:00.000Z",
    "std": "2025-06-05T22:00:00.000Z",
  },
}
`;

exports[`Flights Endpoint > POST /flights > should create a flight with valid data 1`] = `
{
  "aircraft": "CSTRC",
  "departure": "LPPD",
  "destination": "LPLA",
  "flightNumber": "AVIO205",
  "id": "6841cade4cace03b8f75235b",
  "schedule": {
    "sta": "2025-06-05T22:00:00.000Z",
    "std": "2025-06-05T22:00:00.000Z",
  },
}
`;
